ludwig/models/trainer.py

```diff
@@ -91,183 +100,200 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+
+    decay_rate: float = schema.FloatRange(default=0.96, min=0.0, max=1.0)
+    """TODO: Document parameters. (default: 0.96)."""
+
+    staircase: bool = False
+    """Decays the learning rate at discrete intervals (default: False)."""
+
+    gradient_clipping: Optional[GradientClippingConfig] = GradientClippingDataclassField(default={})
+    """Instance of `ludwig.modules.optimization_modules.GradientClippingConfig` that sets gradient clipping params.
+       (default: `ludwig.modules.optimization_modules.GradientClippingConfig()`)"""
+
+    # TODO(#1673): Need some more logic here for validating against output features
+    validation_field: str = COMBINED
+    """First output feature, by default it is set as the same field of the first output feature (default:
+       `ludwig.constants.COMBINED`)."""
+
+    validation_metric: str = LOSS
+    """Metric used on `validation_field`, set by default to accuracy (default: `ludwig.constants.LOSS`)."""
+
+    learning_rate_warmup_epochs: float = schema.NonNegativeFloat(default=1.0)
+    """Number of epochs to warmup the learning rate for (default: 1.0)."""
+
+
 class Trainer(BaseTrainer):
     """Trainer is a class that trains a model."""
 
+    @staticmethod
+    def get_schema_cls():
+        return TrainerConfig
+
     def __init__(
         self,
         model: ECD,
-        optimizer=None,
-        epochs=100,
-        regularization_lambda=0.0,
-        regularization_type=None,
-        learning_rate=0.001,
-        decay=False,
-        decay_rate=0.96,
-        decay_steps=10000,
-        staircase=False,
-        batch_size=128,
-        eval_batch_size=None,
-        should_shuffle=True,
-        bucketing_field=None,
-        validation_field="combined",
-        validation_metric="loss",
-        early_stop=20,
-        reduce_learning_rate_on_plateau=0,
-        reduce_learning_rate_on_plateau_patience=5,
-        reduce_learning_rate_on_plateau_rate=0.5,
-        reduce_learning_rate_eval_metric=LOSS,
-        reduce_learning_rate_eval_split=TRAINING,
-        increase_batch_size_on_plateau=0,
-        increase_batch_size_on_plateau_patience=5,
-        increase_batch_size_on_plateau_rate=2,
-        increase_batch_size_on_plateau_max=512,
-        increase_batch_size_eval_metric=LOSS,
-        increase_batch_size_eval_split=TRAINING,
-        learning_rate_warmup_epochs=1,
-        resume=False,
-        skip_save_model=False,
-        skip_save_progress=False,
-        skip_save_log=False,
-        callbacks=None,
-        random_seed=default_random_seed,
-        horovod=None,
-        device=None,
+        resume: float = False,
+        skip_save_model: bool = False,
+        skip_save_progress: bool = False,
+        skip_save_log: bool = False,
+        callbacks: List = None,
+        random_seed: float = default_random_seed,
+        horovod: Optional[Dict] = None,
+        device: Optional[str] = None,
+        config: Optional[TrainerConfig] = TrainerConfig(),
```

Comment:
Let's set this to `None` so the caller must pass it explcitly.

Reply:
In description follow-up list